src/components/validations/with-validation.hoc.js

```diff
@@ -83,77 +84,79 @@ const withValidation = (WrappedComponent) => {
       if (!isOnSubmit && this.blockValidation) return new Promise(resolve => resolve(true));
 
       const validationPromises = [];
-      types.forEach((type) => {
+      const typesArr = Array.isArray(types) ? types : [types];
+      typesArr.forEach((type) => {
         const validationPromise = this.runValidation(type);
         if (validationPromise) validationPromises.push(validationPromise);
       });
-      return Promise.all(validationPromises).then((results) => {
-        return !results.includes(false);
-      });
+      return Promise.all(validationPromises).then(this.processValidationResults.bind(this));
     }
 
-    updateValidationStatus(type, newMessage = '') {
-      if (type === 'error') this.updateErrorStatus(newMessage);
-      if (type === 'warning') this.updateWarningStatus(newMessage);
-      if (type === 'info') this.updateInfoStatus(newMessage);
+    processValidationResults(results) {
+      if (results.length === 0) {
+        return true;
+      }
+      const { pass: result, state } = results.reduce((acc, { pass, type, error }) => {
+        return {
+          ...acc,
+          ...(acc.pass && !pass && { pass }),
+          state: { ...acc.state, ...this.updateValidationStatus(type, error) }
+        };
+      }, { pass: true });
```

Comment:
This method is extremely difficult to read.
Could you refactor this method to be more readable?